src/transformers/models/grounding_dino/modeling_grounding_dino.py

```diff
@@ -0,0 +1,3714 @@
+
+            layer_head_mask = head_mask[i] if head_mask is not None else None
+            past_key_value = past_key_values[i] if past_key_values is not None else None
+
+            if self.gradient_checkpointing and self.training:
+                layer_outputs = self._gradient_checkpointing_func(
+                    layer_module.__call__,
+                    hidden_states,
+                    attention_mask,
+                    layer_head_mask,
+                    encoder_hidden_states,
+                    encoder_attention_mask,
+                    past_key_value,
+                    output_attentions,
+                )
+            else:
+                layer_outputs = layer_module(
+                    hidden_states,
+                    attention_mask,
+                    layer_head_mask,
+                    encoder_hidden_states,
+                    encoder_attention_mask,
+                    past_key_value,
+                    output_attentions,
+                )
+
+            hidden_states = layer_outputs[0]
+            if use_cache:
+                next_decoder_cache += (layer_outputs[-1],)
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (layer_outputs[1],)
+                if self.config.add_cross_attention:
+                    all_cross_attentions = all_cross_attentions + (layer_outputs[2],)
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(
+                v
+                for v in [
+                    hidden_states,
+                    next_decoder_cache,
+                    all_hidden_states,
+                    all_self_attentions,
+                    all_cross_attentions,
+                ]
+                if v is not None
+            )
+        return BaseModelOutputWithPastAndCrossAttentions(
+            last_hidden_state=hidden_states,
+            past_key_values=next_decoder_cache,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+            cross_attentions=all_cross_attentions,
+        )
+
+
+# Copied from transformers.models.bert.modeling_bert.BertPooler with Bert->GroundingDinoText
+class GroundingDinoTextPooler(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+        self.activation = nn.Tanh()
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        # We "pool" the model by simply taking the hidden state corresponding
+        # to the first token.
+        first_token_tensor = hidden_states[:, 0]
+        pooled_output = self.dense(first_token_tensor)
+        pooled_output = self.activation(pooled_output)
+        return pooled_output
+
+
+class GroundingDinoTextPrenet(GroundingDinoPreTrainedModel):
```

Comment:
At first, I copied from `Bert`, but had to modify some stuff to avoid errors, but also based myself on `CLIP` which has its vision and text model inheriting from `PreTrainedModel`. Should I switch to `nn.Module`?

Reply:
Yes, based on the rest of this PR it should be `nn.Module`. 

For context, BERT is a very old model in the library and so was added before we had the current patterns. This means some of the implementation is non-standard as it's made to fit the current library whilst being backwards compatible. 

For CLIP, the text and vision models both inherit from `PreTrainedModel` because it's intended that they can both be loaded and used as their own independent models. 

For this model, I don't think people will be doing: 

```py
model = GroundingDinoTextPrenet.from_pretrained(checkpoint)
```

Reply:
The reason it's still useful to inherit from `PreTrainedModel` is to be able to leverage the `get_extended_attention_mask` function.